### PR TITLE
Introducing TupleType and TupleExpression Syntax

### PR DESCRIPTION
**Pull request by VSadov**
==== Adding Syntax nodes and Parser support for Tuples.

Most interesting parts here are:
* the syntax node shape and naming - these are public APIs so we should get the naming right.
NOTE: I expect some churn here so for now I have disabled the public API diagnostics.

* parsing something that starts with "("  is tricky since that can be a start of many constructs - lambdas, casts, parenthesized expressions...
I have updated the lookahead logic that tries to figure what we are dealing with, but it may need some tuning to be more robust.
In particular there could be some error recovery work to see that handling of broken code is sensible.

==== Current assumptions:
1) TupleTypes and TupleExpressions must have 2 or more elements/arguments. However we may have to parse 0 and 1 sized expressions for error recovery reasons. Those are syntax errors.

2) TupleType looks like

(int x, int y)  x = ...

I.E.

  ( type [name], type [name])

We may enforce that all or no elements have names. Not sure yet if that needs to be in parser or binder.

3) TupleExpression looks like

var something = (x: foo(), y: bar);

I.E.

  ( [name:] expr, [name:] expr)

just like with the types we may want to enforce that all or none of the arguments have names. Not sure if that is a syntax or binding error.
It is not a part of this change.

### CODE REVIEW

**Comment by VSadov** [comment]
CC @khyperia @jaredpar  @agocke @MadsTorgersen

**Comment by tmat** [comment]
Are there any potential ambiguities between tuple types and parenthesized/cast expressions? 

Specifically in interactive scenarios where we parse declarations, statements and expressions within the same context.

**Comment by tmat** [comment]
E.g. something like ```(T) F() => 1;```

**Comment by VSadov** [comment]
@tmat - that is the trick. There is already some lookahead needed to figure what we have after "(". With tuples we may need to look at more tokens.
So far it does not look ambiguous since tuples must have 2 or more elements/arguments. 
 (T)  is a type or parenthesized expr, depending on context
 (T, U) is a tuple
 (T, U) => starts lambda. We have to look ahead for =>

**Comment by CyrusNajmabadi** [comment]
Awesome!  What's the thinking on the implementation of this at runtime?  System.Tuple?  Some sort of ValueTuple?  Something like an anonymous type?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -6250,6 +6298,56 @@ private ArrayRankSpecifierSyntax ParseArrayRankSpecifier(bool isArrayCreation, b
...
+            var open = this.EatToken(SyntaxKind.OpenParenToken);
+            var list = _pool.AllocateSeparated<TupleElementSyntax>();
+            try
+            {
+                if (this.CurrentToken.Kind != SyntaxKind.CloseParenToken)
+                {
+                    var element = ParseTupleElement();
+                    list.Add(element);
+
+                    while (this.CurrentToken.Kind == SyntaxKind.CommaToken)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Hand written list?  Why not use the existing list parsing functionality in the parser?

**Comment by VSadov** [comment]
@CyrusNajmabadi  -  Some sort of ValueTuple with optional use-site annotations to specify alternative field names if so desired.

**Pull request merged**